perfkitbenchmarker/packages/cassandra.py

```diff
@@ -0,0 +1,204 @@
+             'num_cpus': vm.num_cpus}
+
+  for config_file in [CASSANDRA_ENV_TEMPLATE, CASSANDRA_YAML_TEMPLATE]:
+    local_path = data.ResourcePath(config_file)
+    remote_path = os.path.join(
+        CASSANDRA_DIR, 'conf',
+        os.path.splitext(os.path.basename(config_file))[0])
+    vm.RenderTemplate(local_path, remote_path, context=context)
+
+  # Set up logging in CASSANDRA_DIR/logs
+  vm.RemoteCommand(
+      'sed -i -e "s,log4j.appender.R.File=.*,'
+      'log4j.appender.R.File={0}/logs/system.log," {1}'.format(
+          CASSANDRA_DIR,
+          os.path.join(CASSANDRA_DIR, 'conf', 'log4j-server.properties')))
+
+
+def Start(vm):
+  """Start Cassandra on a VM.
+
+  Args:
+    vm: The target vm. Should already be configured via 'Configure'.
+  """
+  vm.RemoteCommand(
+      'nohup {0}/bin/cassandra -p "{1}" 1> {2} 2> {3} &'.format(
+          CASSANDRA_DIR, CASSANDRA_PID, CASSANDRA_OUT, CASSANDRA_ERR))
+
+
+def Stop(vm):
+  """Stops Cassandra on VM."""
+  vm.RemoteCommand('kill $(cat {0})'.format(CASSANDRA_PID),
+                   ignore_failure=True)
+
+
+def IsRunning(vm):
+  """Returns a boolean indicating whether Cassandra is running on 'vm'."""
+  cassandra_pid = vm.RemoteCommand(
+      'cat {0} || true'.format(CASSANDRA_PID))[0].strip()
+  if not cassandra_pid:
+    return False
+
+  try:
+    vm.RemoteCommand('kill -0 {0}'.format(cassandra_pid))
+    return True
+  except errors.VmUtil.SshConnectionError:
+    logging.warn('%s: Cassandra is not running. '
+                 'Startup STDOUT:\n%s\n\nSTDERR:\n%s',
+                 vm,
+                 vm.RemoteCommand('cat ' + CASSANDRA_OUT),
+                 vm.RemoteCommand('cat ' + CASSANDRA_ERR))
+    return False
+
+
+def CleanNode(vm):
+  """Remove Cassandra data from 'vm'."""
+  data_path = os.path.join(vm.GetScratchDir(), 'cassandra')
+  vm.RemoteCommand('rm -rf {0}'.format(data_path))
+
+
+def _StartCassandraIfNotRunning(vm):
+  if not IsRunning(vm):
+    logging.info('Retrying starting cassandra on %s', vm)
+    Start(vm)
+
+
+def StartCluster(seed_vm, vms):
+  vm_count = len(vms) + 1
+
+  # Cassandra setup
+  logging.info('Starting seed VM %s', seed_vm)
+  Start(seed_vm)
+  logging.info('Waiting for seed to start')
+  time.sleep(20)
+  for i in xrange(5):
+    if not IsRunning(seed_vm):
+      logging.warn('Seed %s: Cassandra not running yet (try %d). Waiting %d.',
+                   seed_vm, i, SLEEP_BETWEEN_NODE_START)
+      time.sleep(SLEEP_BETWEEN_NODE_START)
+    else:
+      break
+  else:
+    raise ValueError('Cassandra failed to start on seed.')
+
+  if vms:
+    # Start the VMs with a small pause in between each, to allow the node to
+    # join.
+    # Starting Cassandra nodes fails when multiple nodes attempt to join the
+    # cluster concurrently.
+    for i, vm in enumerate(vms):
+      time.sleep(SLEEP_BETWEEN_NODE_START)
+      logging.info('Starting non-seed VM %d/%d.', i + 1, len(vms))
```

Comment:
Cassandra fails to start when multiple VMs are joining the cluster
concurrently (there is a comment about this just above).

On Fri, May 15, 2015 at 3:15 PM, yuyanting <notifications@github.com> wrote:

> In perfkitbenchmarker/packages/cassandra.py
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/268#discussion_r30450472>
> :
>
> > +      logging.warn('Seed %s: Cassandra not running yet (try %d). Waiting %d.',
> > +                   seed_vm, i, SLEEP_BETWEEN_NODE_START)
> > +      time.sleep(SLEEP_BETWEEN_NODE_START)
> > +    else:
> > +      break
> > +  else:
> > +    raise ValueError('Cassandra failed to start on seed.')
> > +
> > +  if vms:
> > +    # Start the VMs with a small pause in between each, to allow the node to
> > +    # join.
> > +    # Starting Cassandra nodes fails when multiple nodes attempt to join the
> > +    # cluster concurrently.
> > +    for i, vm in enumerate(vms):
> > +      time.sleep(SLEEP_BETWEEN_NODE_START)
> > +      logging.info('Starting non-seed VM %d/%d.', i + 1, len(vms))
>
> Any reason you would like to start vms 1 by 1?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/268/files#r30450472>
> .
>



-- 
Connor McCoy | Software Engineer | connormccoy@google.com | 206-876-1657